_posts/2019-11-19-demo-fraud-detection.md

```diff
@@ -0,0 +1,223 @@
+
+On a high level, our main processing pipeline looks like this:
+
+```java
+DataStream<Alert> alerts =
+    transactions
+        .process(new DynamicKeyFunction())
+        .keyBy(/* some key selector */);
+        .process(/* actual calculations and alerting */)
+```
+
+We have previously established that each rule defines a **`groupingKeyNames`** parameter. This parameter specifies which combination of fields will be used for the incoming events' grouping. Each rule might use an arbitrary combination of these fields. At the same time, every incoming event potentially needs to be evaluated against multiple rules. This implies that events might simultaneously need to be present at multiple parallel instances of evaluating operators which correspond to different rules and hence will need to be forked. Ensuring such events dispatching is the purpose of `DynamicKeyFunction()`.
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2019-11-19-demo-fraud-detection/shuffle_function_1.png" width="800px" alt="Figure 3: Forking events with Dynamic Key Function"/>
+<br/>
+<i><small>Figure 3: Forking events with Dynamic Key Function</small></i>
+</center>
+<br/>
+
+ `DynamicKeyFunction` iterates over a set of defined rules and prepares every event to be processed by a `keyBy()` function by extracting required grouping keys:
+
+```java
+public class DynamicKeyFunction
+    extends ProcessFunction<Transaction, Keyed<Transaction, String, Integer>> {
+   ...
+  /* Simplified */
+  List<Rule> rules = /* Rules that are initialized somehow.
+                        Details will be discussed in a future blog post. */;
+
+  @Override
+  public void processElement(
+      Transaction event,
+      Context ctx,
+      Collector<Keyed<Transaction, String, Integer>> out) {
+
+      for (Rule rule :rules) {
+       out.collect(
+           new Keyed<>(
+               event,
+               KeysExtractor.getKey(rule.getGroupingKeyNames(), event),
+               rule.getRuleId()));
+      }
+  }
+  ...
+}
+```
+ `KeysExtractor.getKey()` uses reflection to extract the required values of `groupingKeyNames` fields from events and combines them as a single concatenated String key, e.g `"{beneficiaryId=25;payeeId=12}"`. Flink will calculate a hash of this key and effectively pin processing of this particular combination to a specific server in the cluster. This will allow tracking all transactions between _beneficiary #25_ and _payee #12_ and evaluating defined rules within a desired time window.
```

Comment:
```suggestion
 `KeysExtractor.getKey()` uses reflection to extract the required values of `groupingKeyNames` fields from events and combines them as a single concatenated String key, e.g `"{beneficiaryId=25;payeeId=12}"`. Flink will calculate the hash of this key and assign the processing of this particular combination to a specific server in the cluster. This will allow tracking all transactions between _beneficiary #25_ and _payee #12_ and evaluating defined rules within a desired time window.
```